src/Microsoft.OpenApi/Validations/ValidationRuleSet.cs

```diff
@@ -0,0 +1,149 @@
+            {
+                foreach (ValidationRule rule in rules)
+                {
+                    Add(rule);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the rules in this rule set.
+        /// </summary>
+        public IEnumerable<ValidationRule> Rules
+        {
+            get
+            {
+                return _rules.Values.SelectMany(v => v);
+            }
+        }
+
+        /// <summary>
+        /// Add the new rule into rule set.
+        /// </summary>
+        /// <param name="rule">The rule.</param>
+        public void Add(ValidationRule rule)
+        {
+            IList<ValidationRule> typeRules;
+            if (!_rules.TryGetValue(rule.ElementType, out typeRules))
+            {
+                typeRules = new List<ValidationRule>();
+                _rules[rule.ElementType] = typeRules;
+            }
+
+            if (typeRules.Contains(rule))
+            {
+                throw new OpenApiException(SRResource.Validation_RuleAddTwice);
+            }
+
+            typeRules.Add(rule);
+        }
+
+        /// <summary>
+        /// Get the enumerator.
+        /// </summary>
+        /// <returns>The enumerator.</returns>
+        public IEnumerator<ValidationRule> GetEnumerator()
+        {
+            foreach (List<ValidationRule> ruleList in _rules.Values)
+            {
+                foreach (var rule in ruleList)
+                {
+                    yield return rule;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the enumerator.
+        /// </summary>
+        /// <returns>The enumerator.</returns>
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return this.GetEnumerator();
+        }
+
+        private static ValidationRuleSet BuildDefaultRuleSet()
+        {
+            ValidationRuleSet ruleSet = new ValidationRuleSet();
+
+            IEnumerable<Type> allTypes = typeof(ValidationRuleSet).Assembly.GetTypes().Where(t => t.IsClass && t != typeof(object));
+            Type validationRuleType = typeof(ValidationRule);
+            foreach (Type type in allTypes)
+            {
+                if (!type.GetCustomAttributes(typeof(OpenApiRuleAttribute), false).Any())
+                {
+                    continue;
+                }
+
+                var properties = type.GetProperties(BindingFlags.Static | BindingFlags.Public);
+                foreach (var property in properties)
+                {
+                    if (validationRuleType.IsAssignableFrom(property.PropertyType))
+                    {
+                        var propertyValue = property.GetValue(null); // static property
+                        ValidationRule rule = propertyValue as ValidationRule;
+                        if (rule != null)
+                        {
+                            ruleSet.Add(rule);
+                        }
+                    }
+                }
```

Comment:
It's a little risky to me that we rely on the type and not an attribute assigned to each property.

Reply:
It is okay for now though. I don't have a better way than to tag all properties, which is something we want to avoid.

---
In reply to: [157650194](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157650194) [](ancestors = 157650194)

Reply:
Yes. Basically, that's convention. It's ok if we follow up this convention. 

---
In reply to: [157650606](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157650606) [](ancestors = 157650606,157650194)